MDM docs: End user auth (#13130)
fleetdm/fleet: docs/Using Fleet/MDM-macOS-setup.md

> ## End user authentication

Reviewer: ```suggestion
## End user authentication and EULA
```

Using Fleet, you can require end users to authenticate with your identity provider (IdP) and agree to an end user license agreement (EULA) before they can use their new Mac.

To require end user authentication, we will do the following steps:

1. Connect Fleet to your IdP
2. Upload a EULA to Fleet
3. Enable end user authentication

### Step 1: connect Fleet to your IdP

> In Fleet, you can connect Fleet to your IdP using the Fleet UI or fleetctl command-line tool.

Reviewer 2: Recommend shortening to:
```suggestion
You can connect Fleet to your IdP using the Fleet UI or fleetctl command-line tool.
```

> The Fleet UI method is a good start if you're just getting familiar with Fleet.

Reviewer: I don't think we need to explain this here or the part below.

Author: Good call. Removed it.

The fleetctl CLI method enables managing your IdP credentials in a Git repository. This way you can enforce code review and benefit from Git's change history.

Fleet UI:

1. Head to the **Settings > Integrations > Automatic enrollment** page.

2. Under **End user authentication**, enter your IdP credentials and select **Save**.

fleetctl CLI:

> 1. Create `fleet-config.yaml` file:

Reviewer: ```suggestion
1. Create `fleet-config.yaml` file or add to your existing configuration yml file:
```

Maybe?

Author: I like it.

```yaml
apiVersion: v1
kind: config
spec:
  mdm:
    end_user_authentication:
      identity_provider_name: "Okta"
      entity_id: 123
      issuer_url: "https://example.com"
      metadata_url: "https://example.com"
  ...
```

> 2. Add an `mdm.end_user_authentication` key to your YAML document. This key accepts the above keys for your IdP credentials. 

Reviewer: ```suggestion
2. Fill in the relevant information from your IDP under the `mdm.end_user_authentication` key. 
```

3. Run the fleetctl `apply -f fleet-config.yml` command to add your IdP credentials.

4. Confirm that your IdP credentials were saved by running `fleetctl get config`.

> Learn more about "No team" configuration options [here](./configuration-files/README.md#organization-settings).

Reviewer: Why would I want to learn about no team configuration? Give me some teaser.

Author: This was a copy/paste error. I removed the section about "No team." Thanks for catching

> ### Step 2: upload a EULA to Fleet

Reviewer: Is this optional?

Author: Yeah, it is. We should make that clear.

1. Head to the **Settings > Integrations > Automatic enrollment** page.

> 2. Under **End user license agreement (EULA)**, select **Upload** and choose your EULA.

Reviewer: We only accept PDF right?

Author: Yes but the UI clarifies this so I don't think we need to add it to the docs:
![Screenshot 2023-08-03 at 5 05 12 PM](https://github.com/fleetdm/fleet/assets/47070608/67b038d3-30c6-479f-9ed5-395bf5954c15)

### Step 3: enable end user authentication

You can enable end user authentication using the Fleet UI or fleetctl command-line tool.

Fleet UI:

1. Head to the **Controls > macOS settings > macOS setup > End user authentication** page.

> 2. Choose which team you want to enable end user authentication for to to by selecting the desired team in the teams dropdown in the upper left corner.

Reviewer 2: ```suggestion
2. Choose which team you want to enable end user authentication for by selecting the desired team in the teams dropdown in the upper left corner.
```

Author: Thanks Rachael!

3. Select the **On** checkbox and select **Save**.

fleetctl CLI: 

1. Choose which team you want to enable end user authentication on.

> In this example, we'll enable end user authentication on the "Workstations (canary)" team so that the authentication is only required for hosts that automatically enroll to this team.

Reviewer 2: Indenting a few spaces will align this paragraph with the 1st list item:
```suggestion
   In this example, we'll enable end user authentication on the "Workstations (canary)" team so that the authentication is only required for hosts that automatically enroll to this team.
```

> 3. Create a `workstations-canary-config.yaml` file:

Reviewer 2: ```suggestion
2. Create a `workstations-canary-config.yaml` file:
```

```yaml
apiVersion: v1
kind: team
spec:
  team:
    name: Workstations (canary)
    mdm:
      macos_setup:
        enable_end_user_authentication: true
    ...
```

Learn more about team configurations options [here](./configuration-files/README.md#teams).

If you want to enable authentication on hosts that automatically enroll to "No team," we'll need to create an `fleet-config.yaml` file:

```yaml
apiVersion: v1
kind: config
spec:
  mdm:
    macos_setup:
      enable_end_user_authentication: true
  ...
```

Learn more about "No team" configuration options [here](./configuration-files/README.md#organization-settings).

3. Add an `mdm.macos_setup.enable_end_user_authentication` key to your YAML document. This key accepts a boolean value.

> 4. Run the fleetctl `apply -f workstations-canary-config.yml` command to enable authentication for this team.

Reviewer 2: ```suggestion
4. Run the `fleetctl apply -f workstations-canary-config.yml` command to enable authentication for this team.
```
(clear enough as-is, but this makes it more copy-paste-able)

### Step 3: upload the package to Fleet

Fleet UI:

1. Head to the **Controls > macOS settings > macOS setup > Bootstrap package** page.

> 2. Choose which team you want to add the bootstrap package to to by selecting the desired team in the teams dropdown in the upper left corner.

Reviewer 3: ```suggestion
2. Choose which team you want to add the bootstrap package to by selecting the desired team in the teams dropdown in the upper left corner.
```

### Step 3: upload the package to Fleet

In Fleet, you can upload the package using the Fleet UI or fleetctl command-line tool.

> The Fleet UI method is a good start if you're just getting familiar with Fleet.

Reviewer: Don't thik we need to exlain this here.

Author: Sounds good. I think cutting is makes sense

The fleetctl CLI method enables managing the package in a Git repository. This way you can enforce code review and benefit from Git's change history.

Fleet UI:

1. Head to the **Controls > macOS settings > macOS setup > Bootstrap package** page.

2. Choose which team you want to add the bootstrap package to to by selecting the desired team in the teams dropdown in the upper left corner.

> 3. Select **Upload** and choose your bootstrap package.

Reviewer: Any required formats?

Author: Yes, the package has to be `.pkg`. This is clarified by the UI so I don't think we need to add it to the docs:
![Screenshot 2023-08-03 at 5 06 15 PM](https://github.com/fleetdm/fleet/assets/47070608/74e965ff-5569-41bb-bc42-bb3da3f07022)

User sees an error message if it's not a `.pkg`
